Fail to pass stage1

commit 5979cb8d7a246f2d0818bca972468ab9c8a77460 introduces preliminary support of C99 `_Bool`. However, it fails to pass stage1:
```
$ make 
  SHECC	out/shecc-stage1.elf
Aborted (core dumped)
make: *** [Makefile:80: out/shecc-stage1.elf] Error 134
```

Tested on Ubuntu 20.04.6 LTS.

--- a/src/lexer.c
+++ b/src/lexer.c
@@ -83,16 +83,16 @@ typedef enum {
 char token_str[MAX_TOKEN_LEN];
 token_t next_token;
 char next_char;
-bool skip_newline = true;
+int skip_newline = 1;
 
-bool preproc_match;
+int preproc_match;
 
 /* Point to the first character after where the macro has been called. It is
  * needed when returning from the macro body.
  */
 int macro_return_idx;
 
-bool is_whitespace(char c)
+int is_whitespace(char c)
 {
     return c == ' ' || c == '\t';
 }
@@ -100,54 +100,53 @@ bool is_whitespace(char c)
 char peek_char(int offset);
 
 /* is it backslash-newline? */
-bool is_linebreak(char c)
+int is_linebreak(char c)
 {
     return c == '\\' && peek_char(1) == '\n';
 }
 
-bool is_newline(char c)
+int is_newline(char c)
 {
     return c == '\r' || c == '\n';
 }
 
 /* is it alphabet, number or '_'? */
-bool is_alnum(char c)
+int is_alnum(char c)
 {
     return ((c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') ||
             (c >= '0' && c <= '9') || (c == '_'));
 }
 
-bool is_digit(char c)
+int is_digit(char c)
 {
     return c >= '0' && c <= '9';
 }
 
-bool is_hex(char c)
+int is_hex(char c)
 {
     return ((c >= '0' && c <= '9') || (c >= 'a' && c <= 'f') || c == 'x' ||
             (c >= 'A' && c <= 'F'));
 }
 
-bool is_numeric(char buffer[])
+int is_numeric(char buffer[])
 {
-    bool hex = false;
-    int size = strlen(buffer);
+    int hex = 0, size = strlen(buffer);
 
     if (size > 2)
-        hex = buffer[0] == '0' && buffer[1] == 'x';
+        hex = (buffer[0] == '0' && buffer[1] == 'x') ? 1 : 0;
 
     for (int i = 0; i < size; i++) {
-        if (hex && !is_hex(buffer[i]))
-            return false;
-        if (!hex && !is_digit(buffer[i]))
-            return false;
+        if (hex && (is_hex(buffer[i]) == 0))
+            return 0;
+        if (!hex && (is_digit(buffer[i]) == 0))
+            return 0;
     }
-    return true;
+    return 1;
 }
 
 void skip_whitespace()
 {
-    while (true) {
+    while (1) {
         if (is_linebreak(next_char)) {
             source_idx += 2;
             next_char = SOURCE[source_idx];
@@ -162,10 +161,10 @@ void skip_whitespace()
     }
 }
 
-char read_char(bool is_skip_space)
+char read_char(int is_skip_space)
 {
     next_char = SOURCE[++source_idx];
-    if (is_skip_space)
+    if (is_skip_space == 1)
         skip_whitespace();
     return next_char;
 }
@@ -178,7 +177,7 @@ char peek_char(int offset)
 /* Lex next token and returns its token type. Parameter 'aliasing' is used for
  * disable preprocessor aliasing on identifier tokens.
  */
-token_t lex_token_internal(bool aliasing)
+token_t lex_token_internal(int aliasing)
 {
     token_str[0] = 0;
 
@@ -188,7 +187,7 @@ token_t lex_token_internal(bool aliasing)
 
         do {
             token_str[i++] = next_char;
-        } while (is_alnum(read_char(false)));
+        } while (is_alnum(read_char(0)));
         token_str[i] = 0;
         skip_whitespace();
 
@@ -215,15 +214,15 @@ token_t lex_token_internal(bool aliasing)
 
     /* C-style comments */
     if (next_char == '/') {
-        read_char(false);
+        read_char(0);
         if (next_char == '*') {
             /* in a comment, skip until end */
             do {
-                read_char(false);
+                read_char(0);
                 if (next_char == '*') {
-                    read_char(false);
+                    read_char(0);
                     if (next_char == '/') {
-                        read_char(true);
+                        read_char(1);
                         return lex_token_internal(aliasing);
                     }
                 }
@@ -231,7 +230,7 @@ token_t lex_token_internal(bool aliasing)
         } else {
             /* single '/', predict divide */
             if (next_char == ' ')
-                read_char(true);
+                read_char(1);
             return T_divide;
         }
         /* TODO: check invalid cases */
@@ -242,52 +241,52 @@ token_t lex_token_internal(bool aliasing)
         int i = 0;
         do {
             token_str[i++] = next_char;
-        } while (is_hex(read_char(false)));
+        } while (is_hex(read_char(0)));
         token_str[i] = 0;
         skip_whitespace();
         return T_numeric;
     }
     if (next_char == '(') {
-        read_char(true);
+        read_char(1);
         return T_open_bracket;
     }
     if (next_char == ')') {
-        read_char(true);
+        read_char(1);
         return T_close_bracket;
     }
     if (next_char == '{') {
-        read_char(true);
+        read_char(1);
         return T_open_curly;
     }
     if (next_char == '}') {
-        read_char(true);
+        read_char(1);
         return T_close_curly;
     }
     if (next_char == '[') {
-        read_char(true);
+        read_char(1);
         return T_open_square;
     }
     if (next_char == ']') {
-        read_char(true);
+        read_char(1);
         return T_close_square;
     }
     if (next_char == ',') {
-        read_char(true);
+        read_char(1);
         return T_comma;
     }
     if (next_char == '^') {
-        read_char(true);
+        read_char(1);
         return T_bit_xor;
     }
     if (next_char == '~') {
-        read_char(true);
+        read_char(1);
         return T_bit_not;
     }
     if (next_char == '"') {
         int i = 0;
         int special = 0;
 
-        while ((read_char(false) != '"') || special) {
+        while ((read_char(0) != '"') || special) {
             if ((i > 0) && (token_str[i - 1] == '\\')) {
                 if (next_char == 'n')
                     token_str[i - 1] = '\n';
@@ -312,13 +311,13 @@ token_t lex_token_internal(bool aliasing)
                 special = 0;
         }
         token_str[i] = 0;
-        read_char(true);
+        read_char(1);
         return T_string;
     }
     if (next_char == '\'') {
-        read_char(false);
+        read_char(0);
         if (next_char == '\\') {
-            read_char(false);
+            read_char(0);
             if (next_char == 'n')
                 token_str[0] = '\n';
             else if (next_char == 'r')
@@ -337,86 +336,86 @@ token_t lex_token_internal(bool aliasing)
             token_str[0] = next_char;
         }
         token_str[1] = 0;
-        if (read_char(true) != '\'')
+        if (read_char(0) != '\'')
             abort();
-        read_char(true);
+        read_char(1);
         return T_char;
     }
     if (next_char == '*') {
-        read_char(true);
+        read_char(1);
         return T_asterisk;
     }
     if (next_char == '&') {
-        read_char(false);
+        read_char(0);
         if (next_char == '&') {
-            read_char(true);
+            read_char(1);
             return T_log_and;
         };
         if (next_char == '=') {
-            read_char(true);
+            read_char(1);
             return T_andeq;
         }
         skip_whitespace();
         return T_ampersand;
     }
     if (next_char == '|') {
-        read_char(false);
+        read_char(0);
         if (next_char == '|') {
-            read_char(true);
+            read_char(1);
             return T_log_or;
         };
         if (next_char == '=') {
-            read_char(true);
+            read_char(1);
             return T_oreq;
         }
         skip_whitespace();
         return T_bit_or;
     }
     if (next_char == '<') {
-        read_char(false);
+        read_char(0);
         if (next_char == '=') {
-            read_char(true);
+            read_char(1);
             return T_le;
         };
         if (next_char == '<') {
-            read_char(true);
+            read_char(1);
             return T_lshift;
         };
         skip_whitespace();
         return T_lt;
     }
     if (next_char == '%') {
-        read_char(true);
+        read_char(1);
         return T_mod;
     }
     if (next_char == '>') {
-        read_char(false);
+        read_char(0);
         if (next_char == '=') {
-            read_char(true);
+            read_char(1);
             return T_ge;
         };
         if (next_char == '>') {
-            read_char(true);
+            read_char(1);
             return T_rshift;
         };
         skip_whitespace();
         return T_gt;
     }
     if (next_char == '!') {
-        read_char(false);
+        read_char(0);
         if (next_char == '=') {
-            read_char(true);
+            read_char(1);
             return T_noteq;
         }
         skip_whitespace();
         return T_log_not;
     }
     if (next_char == '.') {
-        read_char(false);
+        read_char(0);
         if (next_char == '.') {
-            read_char(false);
+            read_char(0);
             if (next_char == '.') {
-                read_char(true);
+                read_char(1);
                 return T_elipsis;
             }
             abort();
@@ -425,51 +424,51 @@ token_t lex_token_internal(bool aliasing)
         return T_dot;
     }
     if (next_char == '-') {
-        read_char(true);
+        read_char(0);
         if (next_char == '>') {
-            read_char(true);
+            read_char(1);
             return T_arrow;
         }
         if (next_char == '-') {
-            read_char(true);
+            read_char(1);
             return T_decrement;
         }
         if (next_char == '=') {
-            read_char(true);
+            read_char(1);
             return T_minuseq;
         }
         skip_whitespace();
         return T_minus;
     }
     if (next_char == '+') {
-        read_char(false);
+        read_char(0);
         if (next_char == '+') {
-            read_char(true);
+            read_char(1);
             return T_increment;
         }
         if (next_char == '=') {
-            read_char(true);
+            read_char(1);
             return T_pluseq;
         }
         skip_whitespace();
         return T_plus;
     }
     if (next_char == ';') {
-        read_char(true);
+        read_char(1);
         return T_semicolon;
     }
     if (next_char == '?') {
-        read_char(true);
+        read_char(1);
         return T_question;
     }
     if (next_char == ':') {
-        read_char(true);
+        read_char(1);
         return T_colon;
     }
     if (next_char == '=') {
-        read_char(false);
+        read_char(0);
         if (next_char == '=') {
-            read_char(true);
+            read_char(1);
             return T_eq;
         }
         skip_whitespace();
@@ -481,7 +480,7 @@ token_t lex_token_internal(bool aliasing)
         int i = 0;
         do {
             token_str[i++] = next_char;
-        } while (is_alnum(read_char(false)));
+        } while (is_alnum(read_char(0)));
         token_str[i] = 0;
         skip_whitespace();
 
@@ -519,8 +518,17 @@ token_t lex_token_internal(bool aliasing)
         if (aliasing) {
             alias = find_alias(token_str);
             if (alias) {
-                /* TODO: Need more reliable way to identify the token's type */
-                token_t t = is_numeric(alias) ? T_numeric : T_identifier;
+                /* FIXME: comparison with string "bool" is a temporary hack */
+                token_t t;
+
+                if (is_numeric(alias)) {
+                    t = T_numeric;
+                } else if (!strcmp(alias, "_Bool")) {
+                    t = T_identifier;
+                } else {
+                    t = T_string;
+                }
+
                 strcpy(token_str, alias);
                 return t;
             }
@@ -537,7 +545,7 @@ token_t lex_token_internal(bool aliasing)
             source_idx = macro_return_idx;
             next_char = SOURCE[source_idx];
         } else
-            next_char = read_char(true);
+            next_char = read_char(1);
         return lex_token_internal(aliasing);
     }
 
@@ -555,7 +563,7 @@ token_t lex_token_internal(bool aliasing)
  */
 token_t lex_token()
 {
-    return lex_token_internal(true);
+    return lex_token_internal(1);
 }
 
 /* Skip the content. We only need the index where the macro body begins. */
@@ -564,25 +572,25 @@ void skip_macro_body()
     while (!is_newline(next_char))
         next_token = lex_token();
 
-    skip_newline = true;
+    skip_newline = 1;
     next_token = lex_token();
 }
 
 /* Accepts next token if token types are matched. */
-bool lex_accept_internal(token_t token, bool aliasing)
+int lex_accept_internal(token_t token, int aliasing)
 {
     if (next_token == token) {
         next_token = lex_token_internal(aliasing);
-        return true;
+        return 1;
     }
 
-    return false;
+    return 0;
 }
 
 /* Accepts next token if token types are matched. To disable aliasing on next
  * token, use 'lex_accept_internal'.
  */
-bool lex_accept(token_t token)
+int lex_accept(token_t token)
 {
     return lex_accept_internal(token, 1);
 }
@@ -590,21 +598,21 @@ bool lex_accept(token_t token)
 /* Peeks next token and copy token's literal to value if token types are
  * matched.
  */
-bool lex_peek(token_t token, char *value)
+int lex_peek(token_t token, char *value)
 {
     if (next_token == token) {
         if (!value)
-            return true;
+            return 1;
         strcpy(value, token_str);
-        return true;
+        return 1;
     }
-    return false;
+    return 0;
 }
 
 /* Strictly match next token with given token type and copy token's literal to
  * value.
  */
-void lex_ident_internal(token_t token, char *value, bool aliasing)
+void lex_ident_internal(token_t token, char *value, int aliasing)
 {
     if (next_token != token)
         error("Unexpected token");
@@ -617,11 +625,11 @@ void lex_ident_internal(token_t token, char *value, bool aliasing)
  */
 void lex_ident(token_t token, char *value)
 {
-    lex_ident_internal(token, value, true);
+    lex_ident_internal(token, value, 1);
 }
 
 /* Strictly match next token with given token type. */
-void lex_expect_internal(token_t token, bool aliasing)
+void lex_expect_internal(token_t token, int aliasing)
 {
     if (next_token != token)
         error("Unexpected token");
@@ -633,5 +641,5 @@ void lex_expect_internal(token_t token, bool aliasing)
  */
 void lex_expect(token_t token)
 {
-    lex_expect_internal(token, true);
+    lex_expect_internal(token, 1);
 }

--- a/src/parser.c
+++ b/src/parser.c
@@ -240,7 +240,7 @@ int read_constant_infix_expr(int precedence)
         lhs = read_constant_expr_operand();
     }
 
-    while (true) {
+    while (1) {
         op = get_operator();
         current_precedence = get_operator_prio(op);
 
@@ -334,7 +334,7 @@ void cppd_control_flow_skip_lines()
 void check_def(char *alias)
 {
     if (find_alias(alias))
-        preproc_match = true;
+        preproc_match = 1;
 }
 
 void read_defined_macro()
@@ -352,7 +352,7 @@ void read_defined_macro()
 /* read preprocessor directive at each potential positions: e.g.,
  * global statement / body statement
  */
-bool read_preproc_directive()
+int read_preproc_directive()
 {
     char token[MAX_ID_LEN];
 
@@ -374,13 +374,13 @@ bool read_preproc_directive()
             lex_expect(T_gt);
         }
 
-        return true;
+        return 1;
     }
     if (lex_accept(T_cppd_define)) {
         char alias[MAX_VAR_LEN];
         char value[MAX_VAR_LEN];
 
-        lex_ident_internal(T_identifier, alias, false);
+        lex_ident(T_identifier, alias);
 
         if (lex_peek(T_numeric, value)) {
             lex_expect(T_numeric);
@@ -394,7 +394,7 @@ bool read_preproc_directive()
         } else if (lex_accept(T_open_bracket)) { /* function-like macro */
             macro_t *macro = add_macro(alias);
 
-            skip_newline = false;
+            skip_newline = 0;
             while (lex_peek(T_identifier, alias)) {
                 lex_expect(T_identifier);
                 strcpy(macro->param_defs[macro->num_param_defs++].var_name,
@@ -408,18 +408,18 @@ bool read_preproc_directive()
             skip_macro_body();
         }
 
-        return true;
+        return 1;
     }
     if (lex_peek(T_cppd_undef, token)) {
         char alias[MAX_VAR_LEN];
 
-        lex_expect_internal(T_cppd_undef, false);
+        lex_expect_internal(T_cppd_undef, 0);
         lex_peek(T_identifier, alias);
         lex_expect(T_identifier);
 
         remove_alias(alias);
         remove_macro(alias);
-        return true;
+        return 1;
     }
     if (lex_peek(T_cppd_error, NULL)) {
         int i = 0;
@@ -427,7 +427,7 @@ bool read_preproc_directive()
 
         do {
             error_diagnostic[i++] = next_char;
-        } while (read_char(false) != '\n');
+        } while (read_char(0) != '\n');
         error_diagnostic[i] = 0;
 
         error(error_diagnostic);
@@ -441,14 +441,14 @@ bool read_preproc_directive()
             cppd_control_flow_skip_lines();
         }
 
-        return true;
+        return 1;
     }
     if (lex_accept(T_cppd_elif)) {
         if (preproc_match) {
             while (!lex_peek(T_cppd_endif, NULL)) {
                 next_token = lex_token();
             }
-            return true;
+            return 1;
         }
 
         preproc_match = read_constant_expr() != 0;
@@ -459,7 +459,7 @@ bool read_preproc_directive()
             cppd_control_flow_skip_lines();
         }
 
-        return true;
+        return 1;
     }
     if (lex_accept(T_cppd_else)) {
         /* reach here has 2 possible cases:
@@ -468,32 +468,32 @@ bool read_preproc_directive()
          */
         if (!preproc_match) {
             skip_whitespace();
-            return true;
+            return 1;
         }
 
         cppd_control_flow_skip_lines();
-        return true;
+        return 1;
     }
     if (lex_accept(T_cppd_endif)) {
-        preproc_match = false;
+        preproc_match = 0;
         skip_whitespace();
-        return true;
+        return 1;
     }
-    if (lex_accept_internal(T_cppd_ifdef, false)) {
-        preproc_match = false;
+    if (lex_accept_internal(T_cppd_ifdef, 0)) {
+        preproc_match = 0;
         lex_ident(T_identifier, token);
         check_def(token);
 
         if (preproc_match) {
             skip_whitespace();
-            return true;
+            return 1;
         }
 
         cppd_control_flow_skip_lines();
-        return true;
+        return 1;
     }
 
-    return false;
+    return 0;
 }
 
 void read_parameter_list_decl(func_t *fd, int anon);

